File: course_project/src/main/java/ru/ssau/course_project/service/TaskService.java
package ru.ssau.course_project.service;

import jakarta.persistence.EntityNotFoundException;
import ru.ssau.course_project.entity.dto.TaskDto;
import ru.ssau.course_project.entity.enums.PriorityEnum;

import java.time.DateTimeException;
import java.time.LocalDateTime;
import java.util.List;

public interface TaskService {

    TaskDto create(TaskDto dto) throws EntityNotFoundException, IllegalArgumentException;

    TaskDto update( TaskDto dto) throws EntityNotFoundException;

    void delete(Long id);

    List<TaskDto> findAll();

    TaskDto findById(Long id) throws EntityNotFoundException;

    List<TaskDto> findAllByPriority(PriorityEnum priority);

    List<TaskDto> findAllByCreatedAtBetween(LocalDateTime start, LocalDateTime end) throws DateTimeException;

    List<TaskDto> findAllByCreatedAtBefore(LocalDateTime time);

    List<TaskDto> findAllByCreatedAtAfter(LocalDateTime time);

    List<TaskDto> findAllByStatus_Name(String statusName);

    List<TaskDto> findAllByEmployee_Id(Long id) throws EntityNotFoundException;

    List<TaskDto> findAllByEmployee_IdAndStatus_Name(Long id, String statusName) throws EntityNotFoundException;

    List<TaskDto> findAllBySprint_Id(Long id) throws EntityNotFoundException;

    List<TaskDto> findAllBySprint_IdAndStatus_Name(Long id, String statusName) throws EntityNotFoundException;

    List<TaskDto> findAllByEmployee_IdAndSprint_Id(Long id, Long sprintId) throws EntityNotFoundException;

    List<TaskDto> findAllByEmployeeIdAndProjectId(Long id, Long projectId) throws EntityNotFoundException;

    List<TaskDto> findMyTasks();

    List<TaskDto> findMyTasksByStatus(String statusName);

    List<TaskDto> findMyTasksBySprintId(Long sprintId) throws EntityNotFoundException;

    List<TaskDto> findMyTasksByProjectId(Long projectId) throws EntityNotFoundException;

}

File: course_project/frontend/src/app/services/task.service.ts
import {inject, Injectable} from '@angular/core';
import {Observable} from 'rxjs';
import {HttpClient, HttpParams} from '@angular/common/http';
import {PriorityEnum} from '../enums/priority.enum';
import { Task } from '../models/task';
import {environment} from '../../environments/environment';

@Injectable({
  providedIn: 'root',
})
export class TaskService {
  private http = inject(HttpClient);
  private baseUrl = `${environment.apiUrl}/task`;
  private myTasksUrl = `${environment.apiUrl}/my-tasks`;


  create(task: Task): Observable<Task> {
    return this.http.post<Task>(this.baseUrl, task);
  }


  update(task: Task): Observable<Task> {
    return this.http.put<Task>(this.baseUrl, task);
  }


  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }


  findById(id: number): Observable<Task> {
    return this.http.get<Task>(`${this.baseUrl}/${id}`);
  }


  findAll(
    filters?: {
      priority?: PriorityEnum;
      start?: Date;
      end?: Date;
      status?: string;
    }
  ): Observable<Task[]> {
    let params = new HttpParams();

    if (filters?.priority) {
      params = params.set('priority', filters.priority);
    }

    if (filters?.start) {
      params = params.set('start', this.formatDateTime(filters.start));
    }

    if (filters?.end) {
      params = params.set('end', this.formatDateTime(filters.end));
    }

    if (filters?.status) {
      params = params.set('status', filters.status);
    }

    return this.http.get<Task[]>(`${this.baseUrl}/all`, { params });
  }


  findAllByEmployeeId(employeeId: number, status?: string): Observable<Task[]> {
    let params = new HttpParams();

    if (status) {
      params = params.set('status', status);
    }

    return this.http.get<Task[]>(`${this.baseUrl}/all/employee/${employeeId}`, { params });
  }


  findAllBySprintId(sprintId: number, status?: string): Observable<Task[]> {
    let params = new HttpParams();

    if (status) {
      params = params.set('status', status);
    }

    return this.http.get<Task[]>(`${this.baseUrl}/all/sprint/${sprintId}`, { params });
  }


  findAllByEmployeeIdAndSprintId(employeeId: number, sprintId: number): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.baseUrl}/all/employee/${employeeId}/sprint/${sprintId}`);
  }


  findAllByEmployeeIdAndProjectId(employeeId: number, projectId: number): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.baseUrl}/all/employee/${employeeId}/project/${projectId}`);
  }


  findMyTasks(status?: string): Observable<Task[]> {
    let params = new HttpParams();

    if (status) {
      params = params.set('status', status);
    }

    return this.http.get<Task[]>(this.myTasksUrl, { params });
  }


  findMyTasksBySprintId(sprintId: number): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.myTasksUrl}/sprint/${sprintId}`);
  }


  findMyTasksByProjectId(projectId: number): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.myTasksUrl}/project/${projectId}`);
  }

  changeStatus(id: number, statusName: string): Observable<void> {
    return this.http.put<void>(`${this.myTasksUrl}/${id}/change-status`, statusName)
  }


  private formatDateTime(date: Date): string {
    return date.toISOString();
  }
}
